pinot-query-planner/src/main/java/org/apache/pinot/query/planner/nodes/AbstractStageNode.java

```diff
@@ -0,0 +1,49 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.query.planner.nodes;
+
+import java.util.ArrayList;
+import java.util.List;
+
+
+public abstract class AbstractStageNode implements StageNode {
+
+  protected final String _stageId;
+  protected final List<StageNode> _inputs;
+
+  public AbstractStageNode(String stageId) {
```

Comment:
I think the base class constructor should also have a `RelNode` as the argument ?

We do `queryStagePlanner.makePlan(relRoot)` -> relRoot is the `RelNode` output from `logicalPlanner.findBestExp()`.

So when the `StageNode` is being created, it has to be created from a corresponding calcite logical `RelNode` ?

I think this is true even for ExchangeNode because they will be created from `LogicalExchange`

Reply:
they need not be a 1-to-1 mapping. but that's a good point. if we do split/merge of RelNodes during stage planning. we can easily introduce this in the Rel level (except for the stage-specific logic).

you can think of the StageNode is simply a "serializable, pinot specific" version of the rel node. with stage metadata ingested. 

This is a good follow up question regarding "what serializable format to use". but I dont have a good solution in this PR for addressing it

Reply:
yes please let's consider this during serialization design